services/gundeck/src/Gundeck/Push.hs

```diff
@@ -56,82 +63,198 @@ import qualified System.Logger.Class          as Log
+            mkNotificationTarget :: Recipient -> NotificationTarget
+            mkNotificationTarget r = target (r ^. recipientId)
+                                   & targetClients .~ r ^. recipientClients
+
+    forM_ cassandraTargets $ \(psh, (notif, notifTrgt)) -> unless (psh ^. pushTransient) $
+        Stream.add (ntfId notif) notifTrgt (psh ^. pushPayload)
+          =<< view (options . optSettings . setNotificationTTL)
+
+    -- websockets
+
+    let notifIdMap = Map.fromList $ (\(psh, (notif, _)) -> (ntfId notif, (notif, psh))) <$> targets
+    resp <- mapM (compilePushResp notifIdMap) =<< Web.bulkPush (compilePushReq <$> targets)
+
+    -- native push
+
+    sendNotice <- view (options . optFallback . fbPreferNotice)
+    forM_ resp $ \((notif, psh), alreadySent) -> do
+        natives <- nativeTargets psh alreadySent
+        pushNative sendNotice notif psh natives
+
+
+compilePushReq :: (Push, (Notification, List1 (Recipient, [Presence]))) -> (Notification, [Presence])
+compilePushReq (psh, notifsAndTargets) =
+    notifsAndTargets & _2 %~ (mconcat . fmap compileTargets . toList)
+  where
+    compileTargets :: (Recipient, [Presence]) -> [Presence]
+    compileTargets (rcp, pre) = fmap snd
+                              . filter (uncurry (shouldActuallyPush psh))
+                              $ (rcp,) <$> pre
+
+compilePushResp :: Map.Map NotificationId (Notification, Push)
+                -> (NotificationId, [Presence])
+                -> Gundeck ((Notification, Push), [Presence])
+compilePushResp notifIdMap (notifId, prcs) = (, prcs) <$> lkup
+  where
+    lkup          = maybe (throwM internalError) pure $ Map.lookup notifId notifIdMap
+    internalError = ErrorCall "bulkpush: dangling notificationId in response!"
+        -- TODO: is there a better type for this?  (also search this PR for other uses of ErrorCall!)
+
+
+-- | Look up 'Push' recipients in Redis, construct a notifcation, and return all the data needed for
+-- performing the push action.
+--
+-- 'Recipient' can be turned into 'NotificationTarget' needed for storing in C*; the 'Presence's can
+-- be turned into 'PushTarget' needed by Cannon and in 'nativeTargets' for filtering the devices
+-- that need no native push.
+mkNotificationAndTargets :: Push -> Gundeck (Notification, List1 (Recipient, [Presence]))
+mkNotificationAndTargets psh = (,) <$> mkNotif <*> doCollect
+  where
+    mkNotif :: Gundeck Notification
+    mkNotif = do
+        notifId <- mkNotificationId
+        pure $ Notification notifId (psh ^. pushTransient) (psh ^. pushPayload)
+
+    doCollect :: Gundeck (List1 (Recipient, [Presence]))
+    doCollect = zip1 rcps <$> Presence.listAll (view recipientId <$> rcps)
+      where
+        rcps :: [Recipient]
+        rcps = toList . fromRange $ (psh ^. pushRecipients :: Range 1 1024 (Set.Set Recipient))
+
+        zip1 :: [a] -> [b] -> List1 (a, b)
+        zip1 (x:xs) (y:ys) = list1 (x, y) (zip xs ys)
+        zip1 _ _ = error "impossible"
```

Comment:
I understand why this is impossible but since we're in `Gundeck`, perhaps would be nicer to `throwM` instead of `error`?

Reply:
I was about to disagree, then I noticed that listAll indeed is not guaranteed to return non-empty.  Good catch, thanks!